std/algorithm/comparison.d

```diff
@@ -892,116 +892,367 @@ nothrow pure @safe unittest
+                    {
+                        if (ei == primaryRange.length) { break; }
+                    }
+                    else
+                    {
+                        if (primaryRange.empty) { break; }
+                    }
+
+                    /* compare current element in primary range `primaryRange` with each
+                     * non-primary range `r` in `rs` */
+                    static foreach (idx, r; rs)
+                    {
+                        static if (idx != primaryRangeIndex) // not primary other rnage
+                        {
+                            static if (!isInfinite!(typeof(r))) // `r` is finite
+                            {
+                                static if (hasLength!(typeof(primaryRange)) &&
+                                           hasLength!(typeof(r))) // TODO move after `hasIndexingAndLength`
+                                {
+                                    /* primaryRange and `r` have already been checked
+                                     * above to have equal lengths so safe to
+                                     * skip empty check for `r` */
+                                }
+                                else
+                                {
+                                    if (r.empty) { return false; } // check for premature emptying of `r`
+                                }
+                                static if (hasIndexingAndLength!(typeof(r)))
+                                {
+                                    static if (hasIndexingAndLength!(typeof(primaryRange)))
+                                    {
+                                        if (!binaryFun!(pred)(primaryRange[ei], r[ei])) { return false; }
+                                    }
+                                    else
+                                    {
+                                        if (!binaryFun!(pred)(primaryRange.front, r[ei])) { return false; }
+                                    }
+                                }
+                                else
+                                {
+                                    static if (hasIndexingAndLength!(typeof(primaryRange)))
+                                    {
+                                        if (!binaryFun!(pred)(primaryRange[ei], r.front)) { return false; }
+                                    }
+                                    else
+                                    {
+                                        if (!binaryFun!(pred)(primaryRange.front, r.front)) { return false; }
+                                    }
+                                    r.popFront();
+                                }
+                            }
+                            else    // `r` is infinite
+                            {
+                                if (!binaryFun!(pred)(primaryRange.front, r.front)) { return false; }
```

Comment:
... and now we even have five cases :/

Reply:
Yeah, I agree this is ugly. I just need something that works to refactor from...